drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2020 Broadcom
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <sw_isr_table.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <drivers/interrupt_controller/gic.h>
+
+/* TODO: probe for each implemented rdist base */
+#define GIC_RDIST_BASE  DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 1)
+
+#define GICR_SGIBASE_OFFSET	0x10000
+
+/* Offsets from GICD base or GICR sgi_base */
+#define	GIC_CTRL		0x0
+#define	GIC_TYPER		0x4
+#define	GIC_IIDR		0x8
+#define	GIC_IGROUPR		0x80
+#define	GIC_ISENABLER		0x100
+#define	GIC_ICENABLER		0x180
+#define	GIC_ISPENDR		0x200
+#define	GIC_ICPENDR		0x280
+#define	GIC_ISACTIVER		0x300
+#define	GIC_ICACTIVER		0x380
+#define	GIC_IPRIORITYR		0x400
+#define	GIC_ITARGETSR		0x800
+#define	GIC_ICFGR		0xc00
+#define GIC_IGROUPMODR		0xd00
+#define	GIC_SGIR		0xf00
```

Comment:
As for the distributor, we should try to use the values defined in `include/drivers/interrupt_controller/gic.h` (if there are differences, `gic.h` should be modified accordingly):
https://github.com/zephyrproject-rtos/zephyr/blob/0d2f31f2ee9904fe4d6968748dac4f0d353ca7af/include/drivers/interrupt_controller/gic.h#L38

As for the redistributor, we could potentially add new `GICR_` definitions for the `SGI_base` registers in `gic.h` to make things more clear, but I think simply using `GICD_` definitions should suffice, since the redistributor `SGI_base` registers that are descendants from the distributor registers must be identical to the corresponding distributor registers.

In fact, the index 0 of the distributor interrupt line control registers directly map to the corresponding local core redistributor interrupt line control registers (e.g. `GICD_IPRIORITYR0` is local core `GICR_IPRIORITYR0`).

Reply:
ACK.
I will see what is duplicate.

Reply:
@stephanosio  Thanks for the comments. Please continue your feedback. Are you aware who else is using GICv3 with zephyr. I would like to check if legacy mode is desired. In that case I can make a followup patch and test it on V2.  
 But this PR is blocking our base platform support to get in.  So I would like to keep this PR to GICv3 to have absolutely no impact on existing platforms using GICv2. Once it soaks in mainline and no issues found we can think of potential code sharing between V2 V3.
 In fact I have more fixes on GICv3 itself in Zephyr context.

Reply:
I am not suggesting that we should have full code sharing between GICv2 and GICv3, as in having both implementations `#ifdef`-ed in one file. I agree that GICv[1-2] and GICv[3-4] are too different to do that.

My main concerns are the following:
* Duplicate definitions (mainly distributor registers) and their placement
* Lack of proper redistributor handling (at least, we should establish some base framework to facilitate handling of per-core redistributor; e.g. definitions, macros)

As far as the redistributor interrupt line control registers go, IMHO, it would be cleaner to just use the distributor definitions in the `gic.h` instead of defining (duplicate) `GIC_...` definitions here again (i.e. just use `GICD_ISENABLERn`, `GICD_ICENABLERn`, etc.). The `GIC_` prefix just seems wrong here; if any, they should be called `GIC_D_R_` or some sort of that.

Reply:
> Are you aware who else is using GICv3 with zephyr

Not really; maybe Marvell.

> But this PR is blocking our base platform support to get in. So I would like to keep this PR to GICv3 to have absolutely no impact on existing platforms using GICv2.

As long as we are not modifying the GICv2 implementation (`gic.c`) and/or definitions, we should not break anything there. The impact of re-using/adding a few extra definitions to the `gic.h` should be minimal.

Reply:
> Lack of redistributor handling. 

Totally agree. But it has lot more to handle than just adding multi-instance.  And currently we do not even have infrastructure to test it. I strongly feel the SMP support in GIC is a future work.  May be I can just make the rdist_base as a variable and access using a per core index instead of fixed macro to give an impression that its per core. But that a false sense of support.

> Duplicate definitions 

The duplication is w.r.t existing V2 driver for those register set  https://github.com/zephyrproject-rtos/zephyr/pull/24850#discussion_r417964991 . 
Those set of registers have same offset in both GICD and GICR  only the base varies.  But just using a name GICD or GICR will be bit confusing.  I can use `GICD_GICR_  `

Reply:
> May be I can just make the rdist_base as a variable and access using a per core index instead of fixed macro to give an impression that its per core.

That sounds reasonable (e.g. `GIC_RDIST_BASE(0)`). Maybe something like `/* FIXME: Per-core redistributor access is not implemented */` would help too.

> Those set of registers have same offset in both GICD and GICR only the base varies. But just using a name GICD or GICR will be bit confusing. I can use `GICD_GICR_`

FYI, the Linux implementation just uses `GICD_...` as well, since, for all practical purposes, there is no difference. Moreover, re-distributor is a "distributor" after all 😕 

https://github.com/torvalds/linux/blob/690e2aba7beb1ef06352803bea41a68a3c695015/drivers/irqchip/irq-gic-v3.c#L318-L342

Reply:
Redistributor is a distributor afterall !  Ditto. 
That is the reason I attempted to define these access macros which can be used by GICD/GICR / V3/ V2 mode.
I will make a version with  `GIC_DIST_XXX`  to define the common offsets for GICD and GICR like Linux.

Reply:
> I will make a version with `GIC_DIST_XXX` to define the common offsets for GICD and GICR like Linux.

If you are referring to `GIC_DIST_IPRIORITYRn` and others, the point I wanted to make was that Linux simply uses `GICD_...` (distributor) definitions because that mostly makes sense, and we should be able to get away with the existing `GICD_...` definitions in `gic.h` too, instead of new separate `#define`s for them when they are already defined in `gic.h`.

Reply:
linux  defines are offsets from distributor base  vs  being absolute register addresses as in existing  `gic.h `.    Redistributor also uses them as is.
https://github.com/torvalds/linux/blob/690e2aba7beb1ef06352803bea41a68a3c695015/include/linux/irqchip/arm-gic-v3.h#L229-L234

That is the reason the existing defines are not suitable.  Instead V2 driver can later move to use the APIs/macros and reduce LOC.

Reply:
> That is the reason the existing defines are not suitable. Instead V2 driver can later move to use the APIs/macros and reduce LOC.

You are right, the current Zephyr `GICD_...` definitions are `GIC_DIST_BASE + ...`. Can we change these definitions so they are offset-only and use them here?

If you don't feel like updating the GICv[1-2] driver for that, I will look into converting them.

Reply:
@sandeepbrcm In general, I wish we could use a more elegant approach to accessing the GIC registers (GICC, GICD, GICR); but, at least, for the time being, it seems that is not achievable without going through a major overhaul on how GIC drivers access the GIC registers, and it may be too much to ask for that to be done in this PR.

I will look into reworking them in the future; for now, renaming the `GIC_...` definitions to `GIC_DIST_...` (or something that is not prefixed `GIC_` only) and leaving them inside `intc_gicv3.c` should be enough.

Reply:
@stephanosio  In this version I have done changes in V3 driver only.  We can certainly reuse the
`arm_gic_xxx` apis in `gic_common`    I see only change needed is 
`#define GIC_GET_BASE(intid)   GIC_DIST_BASE`   for that.
Also refactoring to see whats common and whats not.
But that should be another step.